Build issue with NAG 7.1 and MPICH 4.3.0: non-interoperable type LOGICAL

Today I tried to build MPICH 4.3.0 with NAG (following some advice in #5892) and got this:
```
Error: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 2525: BIND(C) function MPIR_REQUEST_IS_COMPLETE_C is of non-interoperable type LOGICAL
Errors in declarations, no further processing for MPI_C_INTERFACE_NOBUF
[NAG Fortran Compiler error termination, 1 error, 5 warnings]
make: *** [Makefile:47647: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.stamp] Error 1
```

So I took a look and I think this:
```fortran
FUNCTION MPIR_Request_is_complete_c(request) &
    bind(C, name="PMPIX_Request_is_complete") result(res)
    USE, intrinsic :: iso_c_binding, ONLY : c_int
    USE :: mpi_c_interface_types, ONLY : c_Request
    IMPLICIT NONE
    INTEGER(c_Request), VALUE, INTENT(in) :: request
    LOGICAL :: res
END FUNCTION MPIR_Request_is_complete_c
```
needs to be:

```fortran
FUNCTION MPIR_Request_is_complete_c(request) &
    bind(C, name="PMPIX_Request_is_complete") result(res)
    USE, intrinsic :: iso_c_binding, ONLY : c_int, c_bool
    USE :: mpi_c_interface_types, ONLY : c_Request
    IMPLICIT NONE
    INTEGER(c_Request), VALUE, INTENT(in) :: request
    LOGICAL(c_bool) :: res
END FUNCTION MPIR_Request_is_complete_c
```

where the `LOGICAL` type is now a `LOGICAL(c_bool)`. I made this edit local to my build and NAG seemed happy with it.

NOTE: NAG still fails later with:

```
make[2]: Entering directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.0/build-nag-7.2.13'
  GEN      lib/libmpifort.la
libtool: warning: '/ford1/local/gcc/gcc-12.1.0/lib64/libatomic.la' seems to be moved
NAG Fortran Compiler Release 7.2(Shin-Urayasu) Build 7213
Option error: Unrecognised option -pthread
make[2]: *** [Makefile:13010: lib/libmpifort.la] Error 2
make[2]: Leaving directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.0/build-nag-7.2.13'
make[1]: *** [Makefile:40263: all-recursive] Error 1
make[1]: Leaving directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.0/build-nag-7.2.13'
make: *** [Makefile:10489: all] Error 2
```

I'm trying to figure out how `-pthread` is being passed to it...not sure.

I just checked and this is still in 4.3.1 sadly:
```
  MOD      src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.stamp
NAG Fortran Compiler Release 7.2(Shin-Urayasu) Build 7236
Warning: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 1907: C_INT explicitly imported into MPIR_AINT_ADD_C but not used
Warning: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 1917: C_INT explicitly imported into MPIR_AINT_DIFF_C but not used
Warning: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 2525: C_INT explicitly imported into MPIR_REQUEST_IS_COMPLETE_C but not used
Warning: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 3134: C_INT explicitly imported into MPIR_WTICK_C but not used
Warning: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 3141: C_INT explicitly imported into MPIR_WTIME_C but not used
Error: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.f90, line 2525: BIND(C) function MPIR_REQUEST_IS_COMPLETE_C is of non-interoperable type LOGICAL
Errors in declarations, no further processing for MPI_C_INTERFACE_NOBUF
[NAG Fortran Compiler error termination, 1 error, 5 warnings]
make: *** [Makefile:47647: src/binding/fortran/use_mpi_f08/mpi_c_interface_nobuf.stamp] Error 1
```

and then it fails again with the `-lpthread` (which seems to echo https://github.com/open-mpi/ompi/issues/12413 in Open MPI which @ggouaillardet has fixed there in a PR; maybe the same fix can be applied here?):

```
make[2]: Entering directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36'
  GEN      lib/libmpifort.la
libtool: warning: '/ford1/local/gcc/gcc-12.1.0/lib64/libatomic.la' seems to be moved
NAG Fortran Compiler Release 7.2(Shin-Urayasu) Build 7236
Option error: Unrecognised option -pthread
make[2]: *** [Makefile:13010: lib/libmpifort.la] Error 2
make[2]: Leaving directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36'
make[1]: *** [Makefile:40263: all-recursive] Error 1
make[1]: Leaving directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36'
make: *** [Makefile:10489: all] Error 2
```

NOTE: I would gladly submit a PR to fix at least the first (logical) issue, but I have absolutely no idea where the code is coming from. I think maybe `maint/local_python/binding_f08.py` but...wow. This code is confusing.

Good news: I can confirm that #7551 by @hzhou fixes the `LOGICAL` issue.

Bad news: `--with-hwloc=embedded` does not help my `-pthread` issue:
```
make[2]: Entering directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36'
/bin/sh ./libtool  --tag=FC   --mode=link nagfor -Isrc/binding/fortran/use_mpi -Isrc/binding/fortran/use_mpi_f08 -mismatch -fpp -O2 -version-info 17:1:5   -o lib/libmpifort.la -rpath /ford1/share/gmao_SIteam/MPI/mpich/4.3.1/nag-7.2.36/lib  lib/libmpi.la lib/libf77_mpi.la  lib/libf90_mpi.la lib/libf08_mpi.la
libtool: warning: '/ford1/local/gcc/gcc-12.1.0/lib64/libatomic.la' seems to be moved
libtool: link: nagfor -Wl,-shared  -PIC  -Wl,-Wl,,--whole-archive lib/.libs/libf77_mpi.a lib/.libs/libf90_mpi.a lib/.libs/libf08_mpi.a -Wl,-Wl,,--no-whole-archive  -Wl,-Wl,,-rpath -Wl,-Wl,,/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36/lib/.libs -Wl,-Wl,,-rpath -Wl,-Wl,,/ford1/local/gcc/gcc-12.1.0/lib/../lib64 -Wl,-Wl,,-rpath -Wl,-Wl,,/ford1/share/gmao_SIteam/MPI/mpich/4.3.1/nag-7.2.36/lib -Wl,-Wl,,-rpath -Wl,-Wl,,/ford1/local/gcc/gcc-12.1.0/lib/../lib64 lib/.libs/libmpi.so -L/ford1/share/gmao_SIteam/MPI/mpich/4.3.1/nag-7.2.36/lib -L/ford1/local/gcc/gcc-12.1.0/lib64 -lm -ludev -lpciaccess -lrdmacm -lefa -lnl-3 -lnl-route-3 -libverbs -luuid -lnuma /ford1/local/gcc/gcc-12.1.0/lib/../lib64/libatomic.so -lpthread -ldl -lrt  -mismatch -O2   -pthread -Wl,-Wl,,-soname -Wl,-Wl,,libmpifort.so.12 -o lib/.libs/libmpifort.so.12.5.1
NAG Fortran Compiler Release 7.2(Shin-Urayasu) Build 7236
Option error: Unrecognised option -pthread
make[2]: *** [Makefile:13010: lib/libmpifort.la] Error 2
make[2]: Leaving directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36'
make[1]: *** [Makefile:40263: all-recursive] Error 1
make[1]: Leaving directory '/ford1/share/gmao_SIteam/MPI/src/mpich-4.3.1/build-nag-7.2.36'
make: *** [Makefile:10489: all] Error 2
```

Thing is, that flag doesn't even appear in my `config.log` so I'm not sure it's mpich's fault, but it's being dragged in from elsewhere.

Maybe I should try clang+NAG?

--- a/maint/local_python/binding_f08.py
+++ b/maint/local_python/binding_f08.py
@@ -862,7 +862,12 @@ def dump_interface_function(func, name, c_name, is_large):
         decl_list.append("INTEGER(c_int) :: ierror")
     else:
         ret = "res"
-        decl_list.append("%s :: res" % f08_mapping[func['return']])
+        if func['return'] == 'LOGICAL':
+            # MPIX_Request_is_complete
+            decl_list.append("LOGICAL(c_bool) :: res")
+            uses['c_bool'] = 1
+        else:
+            decl_list.append("%s :: res" % f08_mapping[func['return']])
 
     # ----
     G.out.append("")

